File: server/php/src/Presentation/Controllers/HostedCheckoutController.php
<?php

namespace OnlinePayments\ExampleApp\Presentation\Controllers;

use Fig\Http\Message\StatusCodeInterface;
use OnlinePayments\ExampleApp\Application\Interfaces\Services\HostedCheckoutServiceInterface;
use OnlinePayments\ExampleApp\Configuration\Exceptions\ValidationException;
use OnlinePayments\ExampleApp\Presentation\Extensions\ResponseExtension;
use OnlinePayments\ExampleApp\Presentation\Mappers\HostedCheckoutMapper;
use OnlinePayments\ExampleApp\Presentation\Models\HostedCheckout\Request;
use OnlinePayments\ExampleApp\Presentation\Validators\HostedCheckout\HostedCheckoutValidator;
use Slim\Psr7\Request as SlimRequest;
use Slim\Psr7\Response as SlimResponse;

final readonly class HostedCheckoutController
{
    public function __construct(
        private HostedCheckoutServiceInterface $hostedCheckoutService,
        private HostedCheckoutValidator        $hostedCheckoutValidator
    ) {}

    /**
     * @throws ValidationException
     */
    public function createHostedCheckoutSessions(SlimRequest $request): SlimResponse
    {
        $presentationRequest = Request::fromRequest($request);

        $this->hostedCheckoutValidator->validate($presentationRequest);

        $responseDto = $this->hostedCheckoutService->createHostedCheckout(HostedCheckoutMapper::map($presentationRequest));

        $presentationResponse = HostedCheckoutMapper::mapResponse($responseDto);

        return ResponseExtension::createSlimResponse($presentationResponse, StatusCodeInterface::STATUS_CREATED);
    }

    public function getPaymentByHostedCheckoutId(SlimRequest $request): SlimResponse
    {
        $id = $request->getAttribute('id');
        $result = $this->hostedCheckoutService->getPaymentByHostedCheckoutId($id);
        $presentationResponse = HostedCheckoutMapper::mapGetPaymentResponse($result);

        return ResponseExtension::createSlimResponse($presentationResponse, StatusCodeInterface::STATUS_OK);
    }
}

File: server/python/app/presentation/controllers/hosted_checkout_controller.py
from fastapi import APIRouter, Depends
from dependency_injector.wiring import inject, Provide
from fastapi import status
from app.application.interfaces.services.hosted_checkout_service_interface import HostedCheckoutServiceInterface
from app.configuration.di_container.di_container import Container
from app.presentation.mappers.hosted_checkout_mapper import HostedCheckoutPresentationMapper
from app.presentation.models.get_payment_by_hosted_checkout_id.response import HostedCheckoutPaymentResponse
from app.presentation.models.hosted_checkout.request import HostedCheckoutRequest
from app.presentation.validators.hosted_checkout.hosted_checkout_validator import HostedCheckoutValidator

router = APIRouter(tags=["Hosted Checkout"])


@router.post("/sessions", status_code=status.HTTP_201_CREATED)
@inject
def create_hosted_checkout_session(
        request: HostedCheckoutRequest,
        service: HostedCheckoutServiceInterface = Depends(Provide[Container.hosted_checkout_service])
):
    HostedCheckoutValidator().validate(request)

    return HostedCheckoutPresentationMapper.from_dto(
        service.create_hosted_checkout(HostedCheckoutPresentationMapper.to_dto(request)))


@router.get("/sessions/{id}", response_model=HostedCheckoutPaymentResponse, status_code=status.HTTP_200_OK)
@inject
def get_payment_by_hosted_checkout_id(
        id: str,
        service: HostedCheckoutServiceInterface = Depends(Provide[Container.hosted_checkout_service])
):
    return HostedCheckoutPresentationMapper.from_get_payment_dto(service.get_payment_by_hosted_checkout_id(id))
